# Games 
## Games.01: Handsimulation: MInimax und alpha-beta-Pruning

![Spielbaum](spielbaum.png)

1. Geben Sie für den Spielbaum die Minimax-Bewertungen an. 

| Knoten | Bewertung | 
|--------|-----------|
|   A    |     3     |
|   B    |     3     |
|   C    |     2     |
|   D    |     1     |
|   E    |     9     |
|   F    |     2     |
|   G    |     6     |

2. Markieren Sie die Kanten, die bei alpha-beta-Pruning nicht mehr untersucht werden würden, d.h. wo Pruning stattfinden würde. Geben Sie für jeden Knoten die (sich ändernden) α - und β -Werte an.

| Knoten | Bewertung | Alpha / Beta | 
|--------|-----------|--------------|
|        |           |  $-\infty$ $\infty$ | 
|   B    |     3     |  $-\infty$  3 | 
|   E    |     9     |  9, 9 |
|   F    |     2     |  9, 9 |
| ~~G~~  |     6     |  xxx |
|   C    |     2     |  9, 2 | 
|   D    |     1     |  9, 1 |
|   A    |     3     |  9, 1 |

3. Können die Knoten derart geordnet werden, dass alpha-beta-Pruning eine größere Anzahl von Zweigen abschneidet? Wenn ja, geben Sie eine solche Ordnung an. Wenn nein, begründen Sie Ihre Antwort

Ja die Knoten können besser angeordnet werden. 

So können E und F getauscht werden, damit E auch Eleminiert wird. Weitere Verbesserungen habe ich nicht gefunden. 

# Games.02 Optimale Spiele: Minimax und alpha-beta-Pruning

1. Implementieren Sie den Minimax-Algorithmus (wie in der VL besprochen) am Beispiel Tic Tac Toe in einer Sprache Ihrer Wahl.

**Was brauche ich dafür?**
- Zwei Spieler Min und Max 
    - Die Spieler müssen Züge tätigen können 
- Eine Bewertung des Endzustands eines Spielfelds 
    - +1 wenn Max gewinnt
    - -1 wenn Min gewinnt
    - 0 wenn Unentschieden
- Suchbaum muss erstellt werden
- 

Wie modelliere ich den Suchbaum geeignet?

In [ ]:
import sys 

class Node: 
    next_nodes = []
    value = -2
    spielfeld: list
    player: str
    move: int

    def __init__(self, spielfeld = [' ']*9,  player = 'X', move=-1):
        self.spielfeld = spielfeld
        self.player = player
        self.move = move

    def change_player(self):
        return 'O' if self.player == 'X' else 'X'

    def next_nodes(self) -> list:
        childs = []
        for i in range(9):
            if self.spielfeld[i] == ' ':
                new_spielfeld = self.spielfeld.copy()
                new_spielfeld[i] = self.player
                new_child = Node(spielfeld=new_spielfeld, player=self.change_player())
                childs.append((new_child, self.player+str(i)))                
        return childs


    def __str__(self):
        return f"""Spielfeld:
                 {self.spielfeld[0]} | {self.spielfeld[1]} | {self.spielfeld[2]}
                ---|---|---
                 {self.spielfeld[3]} | {self.spielfeld[4]} | {self.spielfeld[3]}
                ---|---|---
                 {self.spielfeld[4]} | {self.spielfeld[5]} | {self.spielfeld[6]}"""

class Tiktaktoe: 
    spielfeld = [' ']*9
    searchtree = Node()


    def __init__(self):
        pass

    def __str__(self):
        return f"""Spielfeld:
                 {self.spielfeld[0]} | {self.spielfeld[1]} | {self.spielfeld[2]}
                ---|---|---
                 {self.spielfeld[3]} | {self.spielfeld[4]} | {self.spielfeld[3]}
                ---|---|---
                 {self.spielfeld[4]} | {self.spielfeld[5]} | {self.spielfeld[6]}"""



    def utility(self):
        pass
        
    def Utility(self, node: Node) -> int:
        win_positions = [
        (0, 1, 2), (3, 4, 5), (6, 7, 8),  # Reihen
        (0, 3, 6), (1, 4, 7), (2, 5, 8),  # Spalten
        (0, 4, 8), (2, 4, 6)              # Diagonalen
        ]
        for a, b, c in win_positions:
            if node.spielfeld[a] != " " and node.spielfeld[a] == node.spielfeld[b] == node.spielfeld[c]:
                return 1 if node.player == 'X' else -1
        if " " not in node.spielfeld:
            return 0
        return 0
        
    def isTerminal(self, node): 
        if self.Utility(node) != 0: 
            return True

        elif " " not in node.spielfeld:
            return True

        return False


    def maxValue(self, state):
        if self.isTerminal(state): return self.Utility(state)

        self.count_minmax += 1

        v = -999999999999
        for (s, a) in state.next_nodes():
            v = max(v, self.minValue(s))
        return v

    def minValue(self, state):
        if self.isTerminal(state): return self.Utility(state)

        self.count_minmax += 1

        v = 9999999999999
        for (s, a) in state.next_nodes():
            v = min(v, self.maxValue(s))
        return v

    def minimax(self, state):
        val = -999999
        self.count_minmax = 0
        action = None
        for (s, a) in state.next_nodes():
            v = self.minValue(s)
            if (val <= v):
                (val, action) = (v, a)
        return action

    def minimax_alphabeta(self, state):
        val = -999999
        self.count_minmaxalphabeta = 0
        action = None
        for (s, a) in state.next_nodes():
            v = self.minValue_alphabeta(s, -99999999, +99999999)
            if (val <= v):
                (val, action) = (v, a)
        return action

    def maxValue_alpabeta(self, state, alpha, beta):
        if self.isTerminal(state): return self.Utility(state)

        self.count_minmaxalphabeta += 1

        v = -9999999999999999
        for (s, a) in state.next_nodes():
            v = max(v, self.minValue_alphabeta(s, alpha, beta))
            if (v >= beta): return v
            alpha = max(alpha, v)
        return v

    def minValue_alphabeta(self, state, alpha, beta):
        if self.isTerminal(state): return self.Utility(state)

        self.count_minmaxalphabeta += 1

        v = +99999999999999999
        for (s, a) in state.next_nodes():
            v = min(v, self.maxValue_alpabeta(s, alpha, beta))
            if (v <= alpha): return v
            beta = min(beta, v)
        return v

    def run(self):
        pass

In [21]:
t = Tiktaktoe()
print(t.minimax(Node()))
t.count_minmax

X4


294777

In [22]:
print(t.minimax_alphabeta(Node()))
t.count_minmaxalphabeta

X4


11155

Wir können sehen, dass beim Algorithmus mit Alpha Beta pruning eine erhebliche reduktion der zu Untersuchenden Fälle entsteht. 294777 zu 11155. Das ist ungefähr 26x kleiner. 

# Games.03 Minimax vereinfachen

Vereinfachen Sie den Minimax-Algorithmus aus der Vorlesung, indem Sie die Eigenschaft Nullsummenspiel berücksichtigen und die Funktionen Min-Value und Max-Value in eine einzige Funktion ohne explizite Unterscheidung der Spieler zusammenfassen.

Wie? 

Ich muss doch schauen welcher Spieler gerade dran ist um den besten Zug für den jeweiligen Spieler zu bestimmen. 
Aber ich muss nicht wissen welcher Spieler gerade dran ist um das Spielfeld zu bewerten. 
Ich schätze Mal mit "Spieler" sind min und max und nicht der Spieler X und O

In [51]:
class minmax_2(Tiktaktoe):
    def minimax_ez(self, state, alpha, beta, player):
        if self.isTerminal(state): return self.Utility(state)

        v = 0
        for (s, a) in state.next_nodes():
            if player == 'X':
                v = -9999999999999
                v = max(v, self.minimax_ez(s, alpha, beta, 'O'))
                if (v >= beta): return v
                alpha = max(alpha, v)
            elif player == 'O':
                v = +9999999999999
                v = min(v, self.minimax_ez(s, alpha, beta, 'X'))
                if (v <= alpha): return v
                beta = min(beta, v)
            else:
                raise(TypeError('Oh oh'))
        return v

In [52]:
m = minmax_2()
m.minimax_ez(Node(), -99999999, +9999999, 'X')

-1

Ich weiß nicht wovon ich den Beispielbaum nehmen soll. Die Bewertung von dem Minimax Algorithmus und meinem Vereinfachten Algorithmus sollte eigentlich gleich sein. 

# Games.04 Suchtiefe begrenzen 

Eval(s)=3X2(s)+X1(s)−(3O2(s)+O1(s))

Bsp. 1 

|   |   |   |
|---|---|---|
| X | X | X |
| O | O |   |


X_3 = 1 
O_3 = 0
Util = 1 

Bsp. 2

| O | X | O |
|---|---|---|
| X | O | X |
| X | X | O |

X_3 = 0 
O_3 = 1

Util = -1

Bsp. 3 

| X | O | O |
|---|---|---|
| X | X | O |
| X | O |   |

X_3 = 1 
0_3 = 0

Util = 1


Bsp. 4 

| O | X | O |
|---|---|---|
| X | O |   |
| X |   |   |

$Eval(s) = 3*0+1 - (3*1+1) = -3$

Bsp. 5 


| O | O |   |
|---|---|---|
| X | O | X |
| X |   |   |


$Eval(s) = 0*3+1-(3*3-0) = -8$

Bsp. 6

| X | X | O |
|---|---|---|
| X | X | O |
| O | O |

$Eval(s) = 0 - (3*2+0) = -6$


# Games.05: Minimax generalisiert 

![spielbaum_mehrspieler.png](spielbaum_mehrspieler.png)